ibm_cloud_sdk_core/authenticators/iam_authenticator.py

```diff
@@ -0,0 +1,102 @@
+# coding: utf-8
+
+# Copyright 2019 IBM All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from .authenticator import Authenticator
+from ..iam_token_manager import IAMTokenManager
+from ..utils import has_bad_first_or_last_char
+
+
+class IamAuthenticator(Authenticator):
+    authentication_type = 'iam'
+
+    def __init__(self,
+                 apikey,
+                 url=None,
+                 client_id=None,
+                 client_secret=None,
+                 disable_ssl_verification=False,
+                 headers=None,
+                 proxies=None):
+        """
+        :attr str apikey: The apikey
+        :attr str url: The url for authentication
+        :attr str client_id: The client id for rate limiting
+        :attr str client_secret: The client secret for rate limiting
+        :attr bool disable_ssl_verification: enables/ disabled ssl verification
+        :attr dict headers: user-defined headers
+        :attr dict proxies: user-defined proxies
+        """
+        self.token_manager = IAMTokenManager(
+            apikey, url, client_id, client_secret, disable_ssl_verification,
+            headers, proxies)
+        self.validate()
```

Comment:
I wonder if you should validate before you create the token manager, so that you're not passing a null value for apikey if none is given. Probably doesn't make a big difference though